README.rst

Python project *mitschreiben*
-----------------------------

.. image:: https://img.shields.io/codeship/78df14e0-ba83-0137-e4e4-5ec1c160b8c0/master.svg
   :target: https://codeship.com/projects/364816
   :alt: CodeShip

.. image:: https://travis-ci.org/sonntagsgesicht/mitschreiben.svg?branch=master
   :target: https://travis-ci.org/sonntagsgesicht/mitschreiben
   :alt: Travis ci

.. image:: https://img.shields.io/readthedocs/mitschreiben
   :target: http://mitschreiben.readthedocs.io
   :alt: Read the Docs

.. image:: https://img.shields.io/codefactor/grade/github/sonntagsgesicht/mitschreiben/master
   :target: https://www.codefactor.io/repository/github/sonntagsgesicht/mitschreiben
   :alt: CodeFactor Grade

.. image:: https://img.shields.io/codeclimate/maintainability/sonntagsgesicht/mitschreiben
   :target: https://codeclimate.com/github/sonntagsgesicht/mitschreiben/maintainability
   :alt: Code Climate maintainability

.. image:: https://img.shields.io/codecov/c/github/sonntagsgesicht/mitschreiben
   :target: https://codecov.io/gh/sonntagsgesicht/mitschreiben
   :alt: Codecov

.. image:: https://img.shields.io/lgtm/grade/python/g/sonntagsgesicht/mitschreiben.svg
   :target: https://lgtm.com/projects/g/sonntagsgesicht/mitschreiben/context:python/
   :alt: lgtm grade

.. image:: https://img.shields.io/lgtm/alerts/g/sonntagsgesicht/mitschreiben.svg
   :target: https://lgtm.com/projects/g/sonntagsgesicht/mitschreiben/alerts/
   :alt: total lgtm alerts

.. image:: https://img.shields.io/github/license/sonntagsgesicht/mitschreiben
   :target: https://github.com/sonntagsgesicht/mitschreiben/raw/master/LICENSE
   :alt: GitHub

.. image:: https://img.shields.io/github/release/sonntagsgesicht/mitschreiben?label=github
   :target: https://github.com/sonntagsgesicht/mitschreiben/releases
   :alt: GitHub release

.. image:: https://img.shields.io/pypi/v/mitschreiben
   :target: https://pypi.org/project/mitschreiben/
   :alt: PyPI Version

.. image:: https://img.shields.io/pypi/pyversions/mitschreiben
   :target: https://pypi.org/project/mitschreiben/
   :alt: PyPI - Python Version

.. image:: https://img.shields.io/pypi/dm/mitschreiben
   :target: https://pypi.org/project/mitschreiben/
   :alt: PyPI Downloads


mitschreiben (german for 'to take notes') helps recording values during
calculations for later evaluation, e.g. check if the right objects or
values were used or to present the results in structure of tables

It provides a class called Record which is basically used for everything. It grants access to record object, it is used
for the recording and it is a context manager used to trigger whether to record or not.

Example Usage
-------------

In the first ``Record(key = value)`` or ``Record(dictionary)`` is placed where one wants to
record a value. The decorator ``Prefix`` provided by this class is used
to define a key extension under which the recorded value will be stored in the
``Record``. The Prefixes get stacked, so when there is a successive
function call to another function which is prefixed those Prefixes are
concatenated.

.. code:: python

    from mitschreiben import Record

    def magical_stuff_happens(baz, barz):
        return "That's", "great"

    class Foo():

        @Record.Prefix()
        def bar(self, baz, barz)
            some_value1, some_value2 = self.do_something(baz, barz)

            Record(a_key=some_value1, another_key=some_value2)

            return some_value1, some_value2

        @Record.Prefix()
        def do_something(self, baz, barz):

            a_dict = {'again_a_key': baz, 'so_creative': barz}

            Record(a_dict)

            return magical_stuff_happens(baz, barz)

        def __repr__(self):
            return "Foo({})".format(id(self))


Now, since ``Record`` is a contextmanager, the recording will only
happen in such a context. The ``with``-statement activates the recording and returns the current scopes record object
for convenient access. Another thing is, that record level is increased by this statement, leading to record objects
that are only available in that scope. When leaving the ``with`` the outer scopes's record will be extend by the inner
one, by prepending the outer records current prefix stack to each key of the inner one.


.. code:: python

    with Record() as rec:
        foo = Foo()
        foo.do_something("baz", "barz")
        foo.bar("baz","barz")

        print rec.entries


The entries are a dict whose keys are tuples which are the stacked Prefixes. In this way it is possible to determine which method on which object was called, what then led
to successive calls, where in the end a value is recorded. The example above has the following output.

.. code:: python

    {('Foo(42403656).do_something', 'again_a_key'): 'baz', ('Foo(42403656).bar', 'Foo(42403656).do_something', 'again_a_key'): 'baz', ('Foo(42403656).do_something', 'so_creative'): 'barz', ('Foo(42403656).bar', 'a_key'): "That's", ('Foo(42403656).bar', 'another_key'): 'great', ('Foo(42403656).bar', 'Foo(42403656).do_something', 'so_creative'): 'barz'}


Formatting the output
---------------------

The Record can be represented in different formats. The base to this is a *tree of dictionaries*,
implemented by the class ``DictTree`` in ``mitschreiben.formatting``. For the two base outputs however, one
does not need to actually instantiate a ``DictTree`` yourself. The respective methods are

.. code::python

    Record().to_csv_files(PATH)
    Record().to_html_tables(FILENAME, PATH)


Both of these methods produce tables of the output. The idea is that, that certain calculations are made with different
objects, leading to the same keywords. So one obtains a table with row keys (object names) and column keys (the keywords
used to record a value). As the name of the former methods suggests, it produces this tables and writes them as single
.csv files into ``Path``, whereas the latter construct a html document in which one can navigate through the tree structure
and see the tables at those positions where they would be placed in the tree. Those tables would look similar to

.. code:: html

    <div class='panel-elem'><table>
    <tr class='headrow'>
    <th colspan='5'>table</th>
    </tr>
    <tr class='bodyrow'>
    <th> </th>
    <th>a_key</th>
    <th>again_a_key</th>
    <th>another_key</th>
    <th>so_creative</th>
    </tr>
    <tr class='bodyrow'>
    <th>Foo(42403656).bar</th>
    <td>That's</td>
    <td>None</td>
    <td>great</td>
    <td>None</td>
    </tr><tr class='bodyrow'>
    <th>Foo(42403656).do_something</th>
    <td>None</td>
    <td>baz</td>
    <td>None</td>
    <td>barz</td>
    </tr></table></div>
    <div class='panel'>
    <div class='panel-elem'><table>
    <tr class='headrow'>
    <th colspan='2'>table</th>
    </tr>
    <tr class='bodyrow'>
    <th> </th>
    <th>Foo(42403656).do_something</th>
    </tr>
    <tr class='bodyrow'>
    <th>again_a_key</th>
    <td>baz</td>
    </tr><tr class='bodyrow'>
    <th>so_creative</th>
    <td>barz</td>
    </tr></table></div>

Another way would be to work with the ``DictTree`` directly.

.. code:: python

    from mitschreiben.formatting import DictTree

    DT = DictTree(rec.entries)

    tables = DT.make_tables()
    for t in tables:
        print t.pretty_string()
        print

This results in the following output. The first table represents the top
level of the record, whereas the other tabels are named by
*object.function*.

.. code::

                        Values |  a_key | again_a_key | another_key | so_creative
             Foo(42403656).bar | That's |        None |       great |        None
    Foo(42403656).do_something |   None |         baz |        None |        barz

    Foo(42403656).bar
                        Values | again_a_key | so_creative
    Foo(42403656).do_something |         baz |        barz




CHANGES.rst

These changes are listed in decreasing version number order.


Release 0.3
-----------

Release date was |today|


Release 0.2
-----------

December 31th, 2017


Release 0.1
-----------

Release date was July 7th, 2017

test.py
from datetime import datetime
import unittest
import os

from mitschreiben import Record


# dummy functions and classes to test Record and Prefix


def magical_stuff_happens(baz, barz):
    return "That's", "great"


class Foo():
    def __init__(self, name):
        self.name = name

    @Record.Prefix()
    def bar(self, baz, barz):
        some_value1, some_value2 = self.do_something(baz, barz)
        Record(a_key=some_value1, another_key=some_value2)
        return some_value1, some_value2

    @Record.Prefix()
    def do_something(self, baz, barz):
        a_dict = {'again_a_key': baz, 'so_creative': barz}
        Record(a_dict)
        return magical_stuff_happens(baz, barz)

    def __repr__(self):
        return "Foo({})".format(self.name)


def do_stuff():
    with Record() as rec:
        foo = Foo("Rom")
        foo.do_something("baz", "barz")
        foo.bar("baz", "barz")


# Test section


class PrefixTest(unittest.TestCase):
    def setUp(self):
        Record().clear()

    def test_decoration_memory(self):
        do_stuff()
        decorated_functions = {'{}.Foo'.format(__name__): {'do_something', 'bar'}}
        self.assertEqual(Record.Prefix.logged_methods(), decorated_functions)

    def test_Prefix_decorator(self):
        baz = 'baz'
        barz = 'barz'
        name = 'fara'

        with Record() as rec:
            foo = Foo(name)
            foo.do_something(baz, barz)
            foo.bar(baz, barz)

        assumed_record_entries = {('Foo({}).do_something'.format(name), 'again_a_key'): baz,
                                  ('Foo({}).do_something'.format(name), 'so_creative'): barz,
                                  ('Foo({}).bar'.format(name), 'Foo({}).do_something'.format(name), 'again_a_key'): baz,
                                  (
                                  'Foo({}).bar'.format(name), 'Foo({}).do_something'.format(name), 'so_creative'): barz,
                                  ('Foo({}).bar'.format(name), 'a_key'): "That's",
                                  ('Foo({}).bar'.format(name), 'another_key'): "great"}
        self.assertEqual(Record().entries, assumed_record_entries)

    def test_multilevel_record_append_prefix_context(self):
        with Record() as R1:
            with Record().append_prefix('level2'):
                self.assertEqual(Record()._prefix_stack, ['level2'])
                with Record() as R2a:
                    self.assertEqual(Record()._prefix_stack, [])
                    Record(level=Record._record_level, name='R2a')

                    self.assertEqual(Record().entries, {('level',): 2, ('name',): 'R2a'})
                self.assertNotEqual(Record(), R2a)
                self.assertEqual(Record(), R1)
                with Record().append_prefix('another_prefix'):
                    self.assertEqual(Record()._prefix_stack, ['level2', 'another_prefix'])
                    with Record() as R2b:
                        self.assertEqual(Record()._prefix_stack, [])
                        Record(recorded='just something')
                        self.assertEqual(Record().entries, {('recorded',): 'just something'})

                self.assertEqual(Record()._prefix_stack, ['level2'])
            self.assertEqual(Record()._prefix_stack, [])

        assumed_record_entries = {('level2', 'level'): 2,
                                  ('level2', 'name'): 'R2a',
                                  ('level2', 'another_prefix', 'recorded'): 'just something'}
        self.assertEqual(Record().entries, assumed_record_entries)


class RecordTest(unittest.TestCase):
    """Testing Basic Functionality"""

    def setUp(self):
        Record().clear()

    def test_call_to_RecordClass_not_started(self):
        value = 'value'
        a_dict = {'a_key': 'a_value'}

        Record(key=value)
        self.assertEqual(Record().entries, dict())
        Record(a_dict)
        self.assertEqual(Record().entries, dict())
        Record(a_dict, key=value)
        self.assertEqual(Record().entries, dict())

    def test_call_to_RecordClass_start_stop(self):
        value = 'value'

        a_dict = {'a_key': 'a_value'}
        b_dict = {'b_key': 'b_value'}

        Record().start()
        Record(key=value)
        self.assertEqual(Record().entries, {('key',): 'value'})
        Record(a_dict)
        self.assertEqual(Record().entries, {('key',): 'value', ('a_key',): 'a_value'})
        Record().stop()
        Record(b_dict, INT=12345)
        self.assertEqual(Record().entries, {('key',): 'value', ('a_key',): 'a_value'})

    def test_Record_as_context(self):
        a_dict = {'a_key': 'a_value'}
        Record(a_dict, key='value')
        self.assertEqual(Record().entries, dict())

        with Record() as R:
            Record(a_dict)
            self.assertEqual(Record().entries, {('a_key',): 'a_value'})
            self.assertEqual(R, Record())

        self.assertEqual(R.entries, {('a_key',): 'a_value'})

    def test_record_of_RecordObject(self):
        value = 'value'

        a_dict = {'a_key': 'a_value'}
        b_dict = {'b_key': 'b_value'}

        R = Record()
        R._record(key=value)
        self.assertEqual(Record().entries, {('key',): 'value'})
        R._record(a_dict)
        self.assertEqual(Record().entries, {('key',): 'value', ('a_key',): 'a_value'})
        R._record(b_dict, INT=12345)
        self.assertEqual(Record().entries,
                         {('key',): 'value', ('a_key',): 'a_value', ('b_key',): 'b_value', ('INT',): 12345})

    def test_multilevel_record_context(self):
        R1 = Record()
        R2 = Record()
        self.assertEqual(R1, R2)
        with Record() as R1:
            with Record() as R2a:
                self.assertNotEqual(R1, R2a)
                self.assertEqual(Record(), R2a)
            with Record() as R2b:
                self.assertNotEqual(R1, R2b)
                self.assertEqual(Record(), R2b)
            self.assertNotEqual(R2a, R2b)


if __name__ == "__main__":
    import sys

    start_time = datetime.now()

    print('')
    print('======================================================================')
    print('')
    print('run %s' % __file__)
    print('in %s' % os.getcwd())
    print('started  at %s' % str(start_time))
    print('')
    print('----------------------------------------------------------------------')
    print('')

    suite = unittest.TestLoader().loadTestsFromModule(__import__("__main__"))
    testrunner = unittest.TextTestRunner(stream=sys.stdout, descriptions=2, verbosity=2)
    testrunner.run(suite)

    print('')
    print('======================================================================')
    print('')
    print('ran %s' % __file__)
    print('in %s' % os.getcwd())
    print('started  at %s' % str(start_time))
    print('finished at %s' % str(datetime.now()))
    print('')
    print('----------------------------------------------------------------------')
    print('')


test/unittests.py
# -*- coding: utf-8 -*-

# mitschreiben
# ------------
# Python library supplying a tool to record values during calculations
# 
# Author:   sonntagsgesicht, based on a fork of Deutsche Postbank [pbrisk]
# Version:  0.3, copyright Wednesday, 18 September 2019
# Website:  https://github.com/sonntagsgesicht/mitschreiben
# License:  Apache License 2.0 (see LICENSE file)


from datetime import datetime
import unittest
import os
import sys

sys.path.append('.')
sys.path.append('..')

from mitschreiben import Record


# dummy functions and classes to test Record and Prefix


def magical_stuff_happens(baz, barz):
    return "That's", "great"


class Foo():
    def __init__(self, name):
        self.name = name

    @Record.Prefix()
    def bar(self, baz, barz):
        some_value1, some_value2 = self.do_something(baz, barz)
        Record(a_key=some_value1, another_key=some_value2)
        return some_value1, some_value2

    @Record.Prefix()
    def do_something(self, baz, barz):
        a_dict = {'again_a_key': baz, 'so_creative': barz}
        Record(a_dict)
        return magical_stuff_happens(baz, barz)

    def __repr__(self):
        return "Foo({})".format(self.name)


def do_stuff():
    with Record() as rec:
        foo = Foo("Rom")
        foo.do_something("baz", "barz")
        foo.bar("baz", "barz")


# Test section


class PrefixTest(unittest.TestCase):
    def setUp(self):
        Record().clear()

    def test_decoration_memory(self):
        do_stuff()
        decorated_functions = {'{}.Foo'.format(__name__): {'do_something', 'bar'}}
        self.assertEqual(Record.Prefix.logged_methods(), decorated_functions)

    def test_Prefix_decorator(self):
        baz = 'baz'
        barz = 'barz'
        name = 'fara'

        with Record() as rec:
            foo = Foo(name)
            foo.do_something(baz, barz)
            foo.bar(baz, barz)

        assumed_record_entries = {('Foo({}).do_something'.format(name), 'again_a_key'): baz,
                                  ('Foo({}).do_something'.format(name), 'so_creative'): barz,
                                  ('Foo({}).bar'.format(name), 'Foo({}).do_something'.format(name), 'again_a_key'): baz,
                                  (
                                  'Foo({}).bar'.format(name), 'Foo({}).do_something'.format(name), 'so_creative'): barz,
                                  ('Foo({}).bar'.format(name), 'a_key'): "That's",
                                  ('Foo({}).bar'.format(name), 'another_key'): "great"}
        self.assertEqual(Record().entries, assumed_record_entries)

    def test_multilevel_record_append_prefix_context(self):
        with Record() as R1:
            with Record().append_prefix('level2'):
                self.assertEqual(Record()._prefix_stack, ['level2'])
                with Record() as R2a:
                    self.assertEqual(Record()._prefix_stack, [])
                    Record(level=Record._record_level, name='R2a')

                    self.assertEqual(Record().entries, {('level',): 2, ('name',): 'R2a'})
                self.assertNotEqual(Record(), R2a)
                self.assertEqual(Record(), R1)
                with Record().append_prefix('another_prefix'):
                    self.assertEqual(Record()._prefix_stack, ['level2', 'another_prefix'])
                    with Record() as R2b:
                        self.assertEqual(Record()._prefix_stack, [])
                        Record(recorded='just something')
                        self.assertEqual(Record().entries, {('recorded',): 'just something'})

                self.assertEqual(Record()._prefix_stack, ['level2'])
            self.assertEqual(Record()._prefix_stack, [])

        assumed_record_entries = {('level2', 'level'): 2,
                                  ('level2', 'name'): 'R2a',
                                  ('level2', 'another_prefix', 'recorded'): 'just something'}
        self.assertEqual(Record().entries, assumed_record_entries)


class RecordTest(unittest.TestCase):
    """Testing Basic Functionality"""

    def setUp(self):
        Record().clear()

    def test_call_to_RecordClass_not_started(self):
        value = 'value'
        a_dict = {'a_key': 'a_value'}

        Record(key=value)
        self.assertEqual(Record().entries, dict())
        Record(a_dict)
        self.assertEqual(Record().entries, dict())
        Record(a_dict, key=value)
        self.assertEqual(Record().entries, dict())

    def test_call_to_RecordClass_start_stop(self):
        value = 'value'

        a_dict = {'a_key': 'a_value'}
        b_dict = {'b_key': 'b_value'}

        Record().start()
        Record(key=value)
        self.assertEqual(Record().entries, {('key',): 'value'})
        Record(a_dict)
        self.assertEqual(Record().entries, {('key',): 'value', ('a_key',): 'a_value'})
        Record().stop()
        Record(b_dict, INT=12345)
        self.assertEqual(Record().entries, {('key',): 'value', ('a_key',): 'a_value'})

    def test_Record_as_context(self):
        a_dict = {'a_key': 'a_value'}
        Record(a_dict, key='value')
        self.assertEqual(Record().entries, dict())

        with Record() as R:
            Record(a_dict)
            self.assertEqual(Record().entries, {('a_key',): 'a_value'})
            self.assertEqual(R, Record())

        self.assertEqual(R.entries, {('a_key',): 'a_value'})

    def test_record_of_RecordObject(self):
        value = 'value'

        a_dict = {'a_key': 'a_value'}
        b_dict = {'b_key': 'b_value'}

        R = Record()
        R._record(key=value)
        self.assertEqual(Record().entries, {('key',): 'value'})
        R._record(a_dict)
        self.assertEqual(Record().entries, {('key',): 'value', ('a_key',): 'a_value'})
        R._record(b_dict, INT=12345)
        self.assertEqual(Record().entries,
                         {('key',): 'value', ('a_key',): 'a_value', ('b_key',): 'b_value', ('INT',): 12345})

    def test_multilevel_record_context(self):
        R1 = Record()
        R2 = Record()
        self.assertEqual(R1, R2)
        with Record() as R1:
            with Record() as R2a:
                self.assertNotEqual(R1, R2a)
                self.assertEqual(Record(), R2a)
            with Record() as R2b:
                self.assertNotEqual(R1, R2b)
                self.assertEqual(Record(), R2b)
            self.assertNotEqual(R2a, R2b)


if __name__ == "__main__":
    import sys

    start_time = datetime.now()

    print('')
    print('======================================================================')
    print('')
    print(('run %s' % __file__))
    print(('in %s' % os.getcwd()))
    print(('started  at %s' % str(start_time)))
    print('')
    print('----------------------------------------------------------------------')
    print('')

    suite = unittest.TestLoader().loadTestsFromModule(__import__("__main__"))
    testrunner = unittest.TextTestRunner(stream=sys.stdout, descriptions=2, verbosity=2)
    testrunner.run(suite)

    print('')
    print('======================================================================')
    print('')
    print(('ran %s' % __file__))
    print(('in %s' % os.getcwd()))
    print(('started  at %s' % str(start_time)))
    print(('finished at %s' % str(datetime.now())))
    print('')
    print('----------------------------------------------------------------------')
    print('')


mitschreiben/table.py
# -*- coding: utf-8 -*-

# mitschreiben
# ------------
# Python library supplying a tool to record values during calculations
# 
# Author:   sonntagsgesicht, based on a fork of Deutsche Postbank [pbrisk]
# Version:  0.3, copyright Wednesday, 18 September 2019
# Website:  https://github.com/sonntagsgesicht/mitschreiben
# License:  Apache License 2.0 (see LICENSE file)


class Table(object):
    """
        A table consist of columns and rows. Each entry has a row and a column key.
    """

    def __init__(self, default_value=None, name=None, left_upper=None):
        self.row_keys = []
        self.col_keys = []
        self._values = {}
        self._default_value = default_value
        self.name = name
        self.left_upper = left_upper

    @property
    def rows_count(self):
        return len(self.row_keys)

    @property
    def cols_count(self):
        return len(self.col_keys)

    def is_empty(self):
        return len(self.row_keys) == 0 or len(self.col_keys) == 0

    def transpose(self):
        ret = Table(name=self.name, left_upper=self.left_upper, default_value=self._default_value)
        for row_key in self.row_keys:
            for col_key in self.col_keys:
                ret.append(col_key, row_key, self.get(row_key, col_key))
        return ret


    def append(self, row_key, col_key, value):
        col_dict = self._values.get(row_key, None)
        if col_dict is None:
            self._values[row_key] = {col_key: value}
            self.row_keys.append(row_key)
        else:
            col_dict[col_key] = value
        if col_key not in self.col_keys:
            self.col_keys.append(col_key)

    def append_row(self, row_key, row):
        for col_key, value in list(row.items()):
            self.append(row_key, col_key, value)

    def get(self, row_key, col_key):
        row_dict = self._values.get(row_key, None)
        if row_dict is None:
            return self._default_value
        ret = row_dict.get(col_key, self._default_value)
        return ret

    def __call__(self, row_key, col_key):
        return self.get(row_key, col_key)

    def get_row(self, row_key):
        ret = {}
        for col_key in self.col_keys:
            ret[col_key] = self.get(row_key, col_key)
        return ret

    def get_row_list(self, row_key, col_keys):
        """
        returns the values of the row:row_key for all col_keys.
        """
        return [self(row_key, col_key) for col_key in col_keys]

    def get_column(self, col_key):
        ret = {}
        for row_key in self.row_keys:
            ret[row_key] = self.get(row_key, col_key)
        return ret

    def get_default(self):
        return self._default_value

    def sort(self, row_compare=None, column_compare=None):
        ret = Table(name=self.name, left_upper=self.left_upper, default_value=self._default_value)
        sortrow_keys = sorted(self.row_keys, key=row_compare)
        sortcol_keys = sorted(self.col_keys, key=column_compare)
        for row_key in sortrow_keys:
            for col_key in sortcol_keys:
                ret.append(row_key, col_key, self.get(row_key, col_key))
        return ret

    def to_csv(self, leftUpper=None, tabName=None, separator=';'):
        repr = list()
        col_keys = sorted(self.col_keys)
        repr.append(""+separator+separator.join(col_keys))
        for key in self.row_keys:
            elements = [key]+self.get_row_list(key, col_keys)
            elements = list(map(str, elements))
            repr.append(separator.join(elements))
        return "\n".join(repr)

    def pretty_string(self, leftUpper=None, tabName=None, separator=" | "):
        representation = []
        name = tabName if tabName is not None else self.name
        if name is not None:
            representation.append(name)

        if len(self.row_keys) == 0 or len(self.col_keys) == 0:
            return '\n'.join(representation + ['Empty Table'])
        tab = self
        line = leftUpper if leftUpper is not None else (self.left_upper if self.left_upper is not None else '')
        max_row_key_width = max(list(map(len, list(map(str,tab.row_keys)) + [line])))
        line = "{text:>{len}}".format(text=line, len=max_row_key_width)
        max_colkey_width = dict()
        for col_key in tab.col_keys:
            max_colkey_width[col_key] = max([len(str(x)) for x in list(tab.get_column(col_key).values()) + [col_key]])
        for col_key in tab.col_keys:
            line += "{sep}{col:>{width}}".format(sep=separator, col=col_key, width=max_colkey_width[col_key])
        representation.append(line)
        for row_key in tab.row_keys:
            line = "{ROW:>{len}}".format(ROW=row_key, len=max_row_key_width)
            for col_key in tab.col_keys:
                element = str(tab.get(row_key, col_key))
                line += "{sep}{ELEMENT:>{width}}".format(ELEMENT=element, sep = separator, width = max_colkey_width[col_key])
            representation.append(line)
        return '\n'.join(representation)

    def to_html(self):
        rows = self.to_nested_list()
        s = "<table>\n"
        s += "<tr class='headrow'>\n"
        s += "<th colspan='{number}'>{tabname}</th>\n".format(number=len(rows[0]), tabname = self.name)
        s += "</tr>\n"
        header = rows[0]
        s += "<tr class='bodyrow'>\n"
        for cell in header:
            s += "<th>{}</th>\n".format(cell)
        s += "</tr>\n"

        for row in rows[1:]:
            s += "<tr class='bodyrow'>\n"
            s += "<th>{}</th>\n".format(row[0])
            for cell in row[1:]:
                s += "<td>{}</td>\n".format(cell)
            s +="</tr>"

        s+="</table>"
        return s

    def to_nested_list(self):
        """returns the table as a nested list rows"""
        header = [k for k in self.col_keys]
        header.insert(0, " ")
        rows = [header]
        for row_key in self.row_keys:
            row = self.get_row_list(row_key, self.col_keys)
            row.insert(0, row_key)
            rows.append(row)
        return rows

    @staticmethod
    def create_from_json_dict(json_dict):
        ret = Table(name=json_dict['TableName'], left_upper=json_dict['LeftUpper'])
        table_rows = json_dict['Table']
        header = table_rows[0][1:]
        first_col = [row[0] for row in table_rows][1:]
        for r in range(len(first_col)):
            row = table_rows[r + 1]
            for c in range(len(header)):
                ret.append(first_col[r], header[c], row[c + 1])
        return ret



mitschreiben/__init__.py
# -*- coding: utf-8 -*-

# mitschreiben
# ------------
# Python library supplying a tool to record values during calculations
# 
# Author:   sonntagsgesicht, based on a fork of Deutsche Postbank [pbrisk]
# Version:  0.3, copyright Wednesday, 18 September 2019
# Website:  https://github.com/sonntagsgesicht/mitschreiben
# License:  Apache License 2.0 (see LICENSE file)


import logging

logging.getLogger(__name__).addHandler(logging.NullHandler())


__doc__ = 'Python library supplying a tool to record values during calculations'
__version__ = '0.3'
__dev_status__ = '4 - Beta'
__date__ = 'Wednesday, 18 September 2019'
__author__ = 'sonntagsgesicht, based on a fork of Deutsche Postbank [pbrisk]'
__email__ = 'sonntagsgesicht@icloud.com'
__url__ = 'https://github.com/sonntagsgesicht/' + __name__
__license__ = 'Apache License 2.0'
__dependencies__ = ('six',)
__dependency_links__ = ()
__data__ = ('html_basics/*',)
__scripts__ = ()


__all__ = ['Record', 'DictTree']

from .recording import Record
from .formatting import DictTree
from .table import Table
